bindings/rust/src/lib.rs

```diff
@@ -193,6 +198,45 @@ impl Instance {
             0: sys::fizzy_execute(self.0.as_ptr(), func_idx, args.as_ptr(), depth),
         }
     }
+
+    /// Find function type for a given index. Must be a valid index otherwise behaviour is undefine.
+    unsafe fn get_function_type(&self, func_idx: u32) -> sys::FizzyFunctionType {
+        let module = self.get_instance_module();
+        sys::fizzy_get_function_type(module, func_idx)
+    }
+
+    /// Execute a given function of `name` with the given values `args`.
+    ///
+    /// An error is returned if the function can not be found, or inappropriate number of arguments are passed.
+    /// However the types of the arguments are not validated.
+    ///
+    /// The `depth` argument sets the initial call depth. Should be set to `0`.
+    pub fn execute(
+        &mut self,
+        name: &str,
+        args: &[Value],
+        depth: i32,
+    ) -> Result<ExecutionResult, ()> {
+        let func_idx = self.find_exported_function_index(&name);
+        if func_idx.is_none() {
+            return Err(());
+        }
+        let func_idx = func_idx.unwrap();
+
+        let func_type = unsafe { self.get_function_type(func_idx) };
+        if func_type.inputs_size != args.len() {
+            return Err(());
+        }
+        // TODO: validate input types too
+
+        let ret = unsafe { self.unsafe_execute(func_idx, args, depth) };
+        // Fail if output is expected, but none was returned.
```

Comment:
It shouldn't happy unless there're bugs in execution, so maybe this should be an assertion?

Reply:
The opposite could also be asserted: function type is void, but something is returned.

Reply:
We should use assertions when there is an API bug, and errors when it is a user validation question. I guess in this case assertion is fine as we get the `func_idx` from the name.

Reply:
I think this should do it?
```
        if !ret.trapped()
        {
           assert!((func_type.output == sys::FizzyValueTypeVoid) == !ret.value().is_some());
        }
```